### PR TITLE
major rewrites to simplify new seasons

### PR DESCRIPTION
**Pull request by khari05**
~~former title: update for 2021-2022~~

~~TODO:~~

~~- [x] fix building on android~~
~~- [x] finish the match breakdown~~
~~- [x] update localization~~
~~- [ ] maybe more descriptive member variable names for FreightFrenzyAllianceDetails?~~

~~I might get busy, here are my current changes.~~

New features:

Should be able to render match breakdowns using only i18n files and JSON schema

Dynamic dark theme w/ easy-dynamic-theme

TODO:
- [x] cleanup
- [x] add in the remaining seasons
- [x] add error catching for some of my precarious assumptions
- [x] Add back the unable to show data page

### CODE REVIEW

**Comment by khari05** [comment]
Hey @ofekashery @theswerd, can you add HE and ES localization?

**Comment by nholo1332** [comment]
I can add the Spanish translations. I'll try to get that done today

**Comment by khari05** [comment]
I'm currently trying to get the app to build on Android since that is the last step before we can release this version

**Comment by ofekashery** [comment]
Thank you so much, @khari05!

**Comment by nholo1332** [comment]
Awesome, thanks! Once we're ready for this build to be released, I can push the iOS version to the AppStore.

**Review by ofekashery — APPROVED** [review]

**Comment by khari05** [comment]
I'm really happy with how straightforward it was to overhaul the most tedious part of updating the app

~~Only needed like 3 pages of infrastructure diagrams~~

**Comment by khari05** [comment]
In my rough testing, [this previous commit](https://github.com/the-orange-alliance/mobile/pull/31/commits/6a64bca9e519708ab8a72df1b86bc6e502f826e7) makes the team info page open in half the time. Here are some numbers I collected from Dart DevTools:

| first request made | last request finishes | total time    | average        |
|--------------------|-----------------------|---------------|----------------|
| 17:29:21.878       | 17:29:30.791          | 9.03 seconds  |                |
| 17:35:57.557       | 17:36:06.513          | 8.956 seconds | 8.993 seconds  |
| 17:39:49.243       | 17:39:53.116          | 3.873 seconds |                |
| 17:46:59.292       | 17:47:03.080          | 3.788 seconds | 3.8305 seconds |

**Comment by khari05** [comment]
Hey @ofekashery can you take a look at this?

File: lib/models/ranking.dart [hunk 1]
```diff
@@ -39,30 +39,31 @@ class Ranking {
   final Team team;
 
   static List<Ranking> allFromResponse(String response) {
-    return jsonDecode(response)
-        .map((obj) => Ranking.fromMap(obj))
-        .toList()
-        .cast<Ranking>();
+    final list = List.from(json.decode(response));
+    list.removeWhere((e) => e == null); // WORKAROUND FOR SKYSTONE
```
Comment (ofekashery) [inline lib/models/ranking.dart]: Should be resolved now.

File: lib/ui/views/team/subpages/team-results.dart [hunk 2]
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
...
   }
 
   Future<List<TeamParticipant>> getTeamParticipants(String teamKey) async {
     List<TeamParticipant> teamParticipants = [];
 
     // Get all the team's rankings
-    List<Ranking> allRankings = await ApiV3().getTeamResults(teamKey, StaticData().seasonKey);
+    List<Ranking> allRankings = await ApiV3().getTeamResults(teamKey, StaticData.seasonKey);
+
+    final teamEvents = await ApiV3().getTeamEvents(teamKey, StaticData.seasonKey);
```
Comment (ofekashery) [inline lib/ui/views/team/subpages/team-results.dart]: Can we use parallel loading for those?

File: lib/ui/views/team/subpages/team-results.dart [hunk 3]
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
...
-        EventParticipant eventParticipant = events[i];
+      List<Future<void>> requests = [];
+
+      for (final event in teamEvents) {
+        EventParticipant eventParticipant = event;
         TeamParticipant teamParticipant = TeamParticipant();
 
         // Get event detail
-        teamParticipant.event = await ApiV3().getEvent(eventParticipant.eventKey);
+        teamParticipant.event =
```
Comment (ofekashery) [inline lib/ui/views/team/subpages/team-results.dart]: Please add to `requests`.

File: lib/ui/views/team/subpages/team-results.dart [hunk 4]
```diff
@@ -95,31 +103,36 @@ class TeamResults extends StatelessWidget {
...
+            await ApiV3().getEvent(eventParticipant.eventKey);
 
         // Get team matches
-        List<Match> teamMatches = [];
-        await ApiV3().getEventMatches(eventParticipant.eventKey).then((matches) {
+        requests.add(ApiV3().getEventMatches(eventParticipant.eventKey).then((matches) {
+          List<Match> teamMatches = [];
           for (int i = 0; i < matches.length; i++) {
             Match match = matches[i];
             for (int i = 0; i < match.participants.length; i++) {
```
Comment (ofekashery) [inline lib/ui/views/team/subpages/team-results.dart]: Can we rewrite it with `.where`?

**Review by ofekashery — CHANGES_REQUESTED** [review]
Thank you @khari05! That's amazing.

File: lib/ui/views/team/subpages/team-results.dart (hunk 4 above)
Comment (khari05) [reply lib/ui/views/team/subpages/team-results.dart]: definitely, I'm surprised I missed that!

**Comment by khari05** [comment]
Using a bit of json schema and helper functions greatly reduces the code we need to write. I think it took me ≈3 hours to add in this past season.
image

**Review by ofekashery — APPROVED** [review]

**Comment by ofekashery** [comment]
Thank you so much @khari05 ❤️

**Pull request merged**